[REVIEW] feat(pubsub) Extended support in subscriber on external Data Source and Information model RT variables

opcua-tsn-team-kalycito opened the pull request:
This PR adds feature on top of #3739 i.e., it extends the support of external data source variable (RT configuration) in subscriber.

jpfr commented on src/pubsub/ua_pubsub_reader.c:
```diff
@@ -1218,6 +1207,58 @@ UA_Server_DataSetReader_addTargetVariables(UA_Server *server, UA_NodeId *parentN
+            for(UA_UInt16 i = 0; i < anzFields; i++) {
+                if(dataSetMsg->data.keyFrameData.dataSetFields[i].hasValue) {
+                    const UA_VariableNode *rtNode = (const UA_VariableNode *)
+                                                     UA_NODESTORE_GET(server, &dataSetReader->subscribedDataSetTarget.targetVariables[i].targetNodeId);
```
Calling `UA_NODESTORE_GET` is where you loose real time guarantees.
Because this requires synchronisation with the main server.

Instead, move the stuff from the valueBackend directly to the PubSub config.
So you don't have to do the lookup in every cycle.

opcua-tsn-team-kalycito replied:
Need some clarifications here. 
I can understand that new struct member UA_DataValue should be created in DSR config so we can prevent using UA_NODESTORE_GET in ua_pubsub_reader.c while receiving. But I could not create UA_DataValue for each fields as there is no DataSetField config in subscriber (it uses fieldmetadata in datasetmetadata) like it is there in publisher.

For this, I have mapped the values directly into the KeyValuePair of FieldMetaData and used this in ua_pubsub_reader.c while receiving. Does this hold good?

jpfr commented on src/pubsub/ua_pubsub.h:
```diff
@@ -200,6 +200,9 @@ typedef struct UA_DataSetReader {
 void UA_Server_DataSetReader_process(UA_Server *server, UA_DataSetReader *dataSetReader, UA_DataSetMessage* dataSetMsg);
 
+/* Process Network Message in RT configuration using DataSetReader */
+void UA_Server_DataSetReader_processRTMessage(UA_Server *server, UA_DataSetReader *dataSetReader);
```
Please use `UA_Server_DataSetReader_process` also for the RT path.

You can split this the RT path out **inside of** UA_Server_DataSetReader_process.
But everything should go through the same entry point for processing.

opcua-tsn-team-kalycito replied:
Done

andreasebner commented on include/open62541/server_pubsub.h:
```diff
@@ -492,6 +492,48 @@ UA_Server_getDataSetWriterConfig(UA_Server *server, const UA_NodeId dsw,
+UA_TargetVariablesSource_clear(UA_TargetVariablesSourceEntry *subscribedDataSetTarget);
+
+/* Return Status Code after creating TargetVariables in Subscriber AddressSpace */
+UA_StatusCode UA_EXPORT
+UA_Server_DataSetReader_createTargetVariables(UA_Server *server, UA_NodeId dataSetReaderIdentifier,
+                                              UA_TargetVariablesSourceEntry* targetVariables);
```
Personally, I would prefer to use the UA_TargetVariablesSource * and a size as input parameter in our API. The **'UA_TargetVariablesSourceEntry** is more internally.

andreasebner commented on src/pubsub/ua_pubsub_reader.c:
```diff
@@ -1240,22 +1284,37 @@ UA_Server_DataSetReader_process(UA_Server *server, UA_DataSetReader *dataSetRead
+            if(rg->config.rtLevel == UA_PUBSUB_RT_FIXED_SIZE) {
+                for(UA_UInt16 i = 0; i < anzFields; i++) {
+                    if(dataSetMsg->data.keyFrameData.dataSetFields[i].hasValue) {
+                        if(dataSetReader->subscribedDataSet.subscribedDataSetTarget.targetVariables[i].attributeId == UA_ATTRIBUTEID_VALUE) {
+                            memcpy((**(dataSetReader->subscribedDataSet.subscribedDataSetTarget.targetVariables[i].externalDataValueSource)).value.data,
+                                   dataSetMsg->data.keyFrameData.dataSetFields[i].value.data,
+                                   dataSetMsg->data.keyFrameData.dataSetFields[i].value.type->memSize);
```
Here we are accessing the 'user memory'. The external data source PR allows the user to register callbacks to get informed if there is read or write access. Currently your implementation doesn't trigger this CB.

jpfr commented:
Looking very good.

As usual, I keep a close eye on public interface definitions.
So I have some wishes there.
Use the following for the new definitions in `server_pubsub.h`.

```c
typedef struct {
    /* Standard-defined FieldTargetDataType */
	UA_FieldTargetDataType targetVariable;

    /* If realtime-handling is required, set this pointer non-NULL and it will be used
     * to memcpy the value instead of using the Write service.
     * If the afterWrite method pointer is set, it will be called after a memcpy update
     * to the value. */
    UA_DataValue **externalDataValue;
    void *externalDataValueContext; /* user-defined pointer */
    void (*afterWrite)(UA_Server *server,
                       const UA_NodeId *dataSetReaderIdentifier,
                       const UA_NodeId *readerGroupIdentifier, 
                       void *externalDataValueContext,
                       const UA_DataValue **externalDataValue);
} UA_FieldTargetVariable;

typedef struct {
    size_t targetVariablesSize;
    UA_FieldTargetVariable *targetVariables;
} UA_TargetVariables;
```

- Reuses standard-defined data types
- Data type names short and to the point
- Allow the user to add a callback method if he wishes to be informed about an incoming value change

Can be merged then.

opcua-tsn-team-kalycito commented:
@jpfr Thanks for restructuring. Completely in-line with the usage of afterWrite callback which is to notify the write to the user in RT path. Can you clarify why do we need an additional pointer "externalDataValueContext" for memcpy as we are already copying the value inside externalDataValue double pointer? Can we use a bool called "rtValueSourceEnabled" for the check instead of a pointer "externalDataValueContext" so that it is understandable to the user?

jpfr commented:
> Can you clarify why do we need an additional pointer "externalDataValueContext" for memcpy as we are already copying the value inside externalDataValue double pointer?

It could be that the user wants to trigger an third-party API when an update is received.
Having a context pointer is a standard pattern used in many places of the stack.
You could rename it to `targetVariableContext` or similar of that is more clear.

>  Can we use a bool called "rtValueSourceEnabled" for the check instead of a pointer "externalDataValueContext" so that it is understandable to the user?

The context can be NULL. If rt is enabled or not should depend on the DataValue double-pointer to be non-null.

jpfr commented:
Thanks for the update.

I had a look at `server_pubsub.h`. The datatype `UA_TargetVariables` was purely a wrapper and not very useful.
See the added commit with some simplifications (but not functional changes).

Now to the problematic part:

In the `UA_DataSetReaderConfig`, the FieldTargetVariables are stored as well.

However,
- this targetVariables in the UA_DataSetReaderConfig are currently not set at all!
- the information is duplicated in the `UA_DataSetReader`. Don't duplicate information!
- changes to the targetVariables are not considered in `UA_Server_DataSetReader_updateConfig`

We should avoid redundant API. It should be enough to set the targetvariables in one place.
If we set the target variables it in the config, then we don't need to have a UA_Server_DataSetReader_createTargetVariables at all!

That is preferrable because it keeps the API surface smaller.

opcua-tsn-team-kalycito commented:
@jpfr Thanks for the comments.

>  the information is duplicated in the UA_DataSetReader. Don't duplicate information!

Yes, you are right, we should avoid redundant information in DataSetReader.

> this targetVariables in the UA_DataSetReaderConfig are currently not set at all!
> changes to the targetVariables are not considered in UA_Server_DataSetReader_updateConfig

From the usage of TargetVariables, I think we don't need definition in DataSetReaderConfig. As TargetVariables are similar to dataSetFields, the current implementation can be modified to use addTargetVariable and removeTargetVariable API in the future to add or remove the variables (similar to DataSetFields) instead of modifying in updateReaderConfig API.

>  If we set the target variables it in the config, then we don't need to have a UA_Server_DataSetReader_createTargetVariables at all!

UA_Server_DataSetReader_createTargetVariables API is required, as in future, we will be using this API to create the user defined Target variable nodes under TargetVariablesType Object (which will be created under DataSetReader Object). 

For now, I have removed the targetVariables from DataSetReaderConfig and modified TSN examples in the new commit. In case of any design change in TargetVariable, can we do that in a separate PR?

andreasebner commented on examples/pubsub/tutorial_pubsub_subscribe.c:
```diff
@@ -136,9 +108,70 @@ addSubscribedVariables (UA_Server *server, UA_NodeId dataSetReaderId) {
+
+    retval |= UA_Server_addDataSetReader(server, readerGroupIdentifier, &readerConfig,
+                                         &readerIdentifier);
+
```
Maybe we can also show the use of our dedicated function 'createTargetVariables()'.

andreasebner commented on include/open62541/server_pubsub.h:
```diff
@@ -492,6 +492,59 @@ UA_Server_getDataSetWriterConfig(UA_Server *server, const UA_NodeId dsw,
+     * to the value. */
+    UA_DataValue **externalDataValue;
+    void *targetVariableContext; /* user-defined pointer */
+    void (*afterWrite)(UA_Server *server,
```
The afterwrite could maybe be part of the external data source itself (more generic).

andreasebner commented on include/open62541/server_pubsub.h:
```diff
@@ -492,6 +492,59 @@ UA_Server_getDataSetWriterConfig(UA_Server *server, const UA_NodeId dsw,
+                       const UA_NodeId *targetVariableIdentifier,
+                       void *targetVariableContext,
+                       UA_DataValue **externalDataValue);
+} UA_FieldTargetVariable;
```
Maybe UA_TargetVariableField would be more clear.

andreasebner commented on src/pubsub/ua_pubsub_reader.c:
```diff
@@ -1091,41 +1134,72 @@ UA_DataSetReader_setPubSubState(UA_Server *server, UA_PubSubState state, UA_Data
- * the given parent node and creates the corresponding data in the
- * targetVariables of the DataSetReader */
-UA_StatusCode
+/* Functionality of addTargetVariables API here is the same with respect to createMirrorVariables.
+ * This API will be used in future to create mirror Variables - TODO */
```
Why don't we call it internally addMirrorVariable?
